internal/cluster/distribution/eks/eksadapter/node_pool_manager.go

```diff
@@ -123,45 +126,91 @@ func generateNodePoolStackName(clusterName string, poolName string) string {
 }
 
 // ListNodePools lists node pools from a cluster.
-func (n nodePoolManager) ListNodePools(ctx context.Context, cluster cluster.Cluster, nodePoolNames []string) ([]eks.NodePool, error) {
-	clusterClient, err := n.dynamicClientFactory.FromSecret(ctx, cluster.ConfigSecretID.String())
+func (n nodePoolManager) ListNodePools(
+	ctx context.Context,
+	c cluster.Cluster,
+	existingNodePools map[string]eks.ExistingNodePool,
+) (nodePools []eks.NodePool, err error) {
+	if c.ConfigSecretID.ResourceID == "" || // Note: cluster is being created or errorred before k8s secret would be available.
+		c.Status == cluster.Deleting {
+		return nil, cluster.NotReadyError{
+			OrganizationID: c.OrganizationID,
+			ID:             c.ID,
+			Name:           c.Name,
+		}
+	}
+
+	clusterClient, err := n.dynamicClientFactory.FromSecret(ctx, c.ConfigSecretID.String())
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed",
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
+			"configSecretID", c.ConfigSecretID.String(),
+		)
 	}
 
 	manager := npls.NewManager(clusterClient, n.namespace)
 	labelSets, err := manager.GetAll(ctx)
 	if err != nil {
 		return nil, errors.WrapWithDetails(err, "retrieving node pool label sets failed",
-			"cluster", cluster,
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
 			"namespace", n.namespace,
 		)
 	}
 
-	awsClient, err := n.awsFactory.New(cluster.OrganizationID, cluster.SecretID.ResourceID, cluster.Location)
+	awsClient, err := n.awsFactory.New(c.OrganizationID, c.SecretID.ResourceID, c.Location)
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating aws factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating aws factory failed",
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
+		)
 	}
 
 	cfClient := n.cloudFormationFactory.New(awsClient)
 	describeStacksInput := cloudformation.DescribeStacksInput{}
-	nodePools := make([]eks.NodePool, 0, len(nodePoolNames))
-	for _, nodePoolName := range nodePoolNames {
-		stackName := generateNodePoolStackName(cluster.Name, nodePoolName)
-		describeStacksInput.StackName = &stackName
+	nodePools = make([]eks.NodePool, 0, len(nodePools))
+	for _, existingNodePool := range existingNodePools {
+		nodePools = append(nodePools, eks.NodePool{
+			Name:   existingNodePool.Name,
+			Labels: labelSets[existingNodePool.Name],
+		})
+		nodePool := &nodePools[len(nodePools)-1]
+
+		stackIdentifier := existingNodePool.StackID
+		if stackIdentifier == "" { // Note: CloudFormation stack creation not started yet.
+			stackIdentifier = generateNodePoolStackName(c.Name, existingNodePool.Name)
```

Comment:
Ideally, we should drop stack name generation at some point which means we won't be able to look it up by name either.

What's our plan for that case?

I could imagine sending a query to the running workflow, but then we need to address it somehow.

Or did I misunderstood something here? The comment indicates that this is a fallback for a non-existing stack, not an old node pool without a stack ID.

Reply:
Your understanding is correct, I haven't thought about stack name elimination yet, thinking about it now.

Reply:
If we have neither stack name nor stack ID yet, then we haven't started creating the stack yet (`CreateStack()` immediately returns a stack ID which is then saved to DB).

In such a case the node pool is either creating or failed with a pre-stack issue (such as validation).
For these we should rely on the DB stored status and status message (those are updated at node pool creation start and pre-stack validation failure) and fill the returned node pool with those values, just like it is done in the `err != nil` handling of the `DescribeStacks()` call in the current implementation.

We could also fill the DB stored creation parameter values of the node pool if we wish, but unless we have an incentive to do so I would avoid that as currently we are only storing node pool creation parameter values for legacy reasons and thus those are up for elimination.
I think a zero valued node pool with a correct `CREATING / ERROR` node pool status is sufficient, we don't actually need to fill the creation parameter values as the node pool is not created yet.

Reply:
We don't necessarily have to store it in the DB. We can send a query to the running cadence workflow, to return some information. It could return the original node pool create request. This sounds like the simplest solution to me at the moment, but we still have to solve:

> then we need to address it somehow

We need to know the process ID (workflow ID) for the create workflow, which we should know (somehow). But keep in mind: we need to do that for cluster creation and node pool creation as well. (Probably need to move node pool creation to child workflows and address them directly with queries)

If this is only an issue in a transient state, I'd say postpone it to a later PR.

Reply:
Yes, it is only an issue in transient state (pre-stack creating and deleting states, but the latter is covered with the stack ID usage).
From the moment of starting stack creation we are going to be able to use the stack ID safely all the time.
So I agree with postponing the decision.

I'm getting used to the idea of accessing the async workflows directly (not against it, just gotta think about its trade-offs more before I can say I'm fine with it).
I guess it's up for debate which is better, to avoid storing such values or to avoid accessing running workflow information directly (with the latter being useful in a lot of other cases as well I feel).
In general I would also prefer derived dynamic state (running workflow details) over stored static state, so I can be easily convinced, just getting used to the idea.

Reply:
Let's open an issue with the details/links to this discussion then.

Reply:
https://github.com/banzaicloud/pipeline/issues/3195